File: scripts/algorithms/Add Two Numbers II/Add Two Numbers II.java
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode res=new ListNode(0);
        ListNode curr=res;
         l1=reverseLinkedList(l1);
         l2=reverseLinkedList(l2);
        int carry=0;
        while(l1!=null||l2!=null||carry==1)
        {
            int sum=0;
            if(l1!=null)
            {
                sum+=l1.val;
                l1=l1.next;
            }
            if(l2!=null)
            {
                sum+=l2.val;
                l2=l2.next;
            }
            sum+=carry;
            carry = sum/10; 
            curr.next= new ListNode(sum % 10); 
            
            curr = curr.next; 
        }
        return reverseLinkedList(res.next);
    }
    public ListNode reverseLinkedList(ListNode head)
    {
        ListNode curr=head;
        ListNode prev=null;
        ListNode next;
        while(curr!=null)
        {
            next=curr.next;
            curr.next=prev;
            prev=curr;
            curr=next;
        }
        return prev;
    }
}

File: scripts/Add Two Numbers II/Add Two Numbers II.cpp
class Solution {
public:
    ListNode* reverseList(ListNode* head){
        if(head == NULL) return head;
        ListNode* ptr, *fptr, *prev;
        ptr = head;
        fptr = ptr->next;
        ptr-> next = NULL;
        while(fptr != NULL){
            prev = ptr;
            ptr= fptr;
            fptr = fptr->next;
            ptr->next = prev;
        }
        return ptr;
    }
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* head = new ListNode();
        ListNode* ptr = head;
        l1 = reverseList(l1);
        l2 = reverseList(l2);
        int carry = 0;
        while(l1 && l2){ 
            ptr->next = new ListNode();
            ptr = ptr->next;
            ptr->val = l1->val + l2->val  + carry;
            carry = ptr->val/10;
            ptr->val = ptr->val % 10;
            l1 = l1->next;
            l2 = l2->next;
        }
        while(l1){ 
            ptr->next = new ListNode();
            ptr = ptr->next;
            ptr->val = l1->val + carry;
            carry = ptr->val/10;
            ptr->val = ptr->val % 10;
            l1 = l1->next;
        }
        while(l2){ 
            ptr->next = new ListNode();
            ptr = ptr->next;
            ptr->val = l2->val  + carry;
            carry = ptr->val/10;
            ptr->val = ptr->val % 10;
            l2 = l2->next;
        }
        if(carry){
            ptr->next = new ListNode();
            ptr = ptr->next;
            ptr->val = carry;
        }
        head = reverseList(head->next);
        return head;
    }
};
